src/session.c

```diff
@@ -436,6 +434,16 @@ static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
 
             pin = (CK_UTF8CHAR_PTR)cb_pin;
             pinlen = cb_pin_len;
+
+            if (p11prov_ctx_cache_pins(session->provctx)) {
+                CK_RV trv;
+                trv = p11prov_slot_set_cached_pin(slot, (const char *)pin);
+                /* not much we can do on failure */
+                if (trv != CKR_OK) {
+                    P11PROV_raise(session->provctx, trv, "Failed to cache pin");
+                }
+            }
+
```

Comment:
My initial though was that we should probably cache the pin only after the login is successful. But the current implementation look like not causing any harm. Maybe just to avoid some noise, when the pin would be cached, but login will fail and on another attempt it will be reported as a bad pin.

Reply:
We should cache only if successful, it is what I intended.
Will fix the code.

Reply:
While looking at this I also realized I did not free the cached_pin in the slot free function, fixed that as well.